[Ho Lanzan] iP

lzan98 opened the pull request:
### DukeDog

> "The best chat bot to ever exist" - Albert Ainstein ([source](https://www.google.com/search?q=The+best+chat+bot+to+ever+exist&sxsrf=APq-WBsEzSRK1xXkab-f9jmGQO2XNXZMGA%3A1643440525396&source=hp&ei=jen0YeiiFumRseMP0oyW6AY&iflsig=AHkkrS4AAAAAYfT3ndbPPmmg0xsPEoIGNmeHUii8g-wt&ved=0ahUKEwioneWFtdb1AhXpSGwGHVKGBW0Q4dUDCA4&uact=5&oq=The+best+chat+bot+to+ever+exist&gs_lcp=Cgdnd3Mtd2l6EANQAFgAYKoEaABwAHgAgAFYiAFYkgEBMZgBAKABAqABAQ&sclient=gws-wiz))

DukePro frees your mind of having to remember things you need to do. It's,
- Text-based
- Easy to use
- <s>Fast</s> _SUPER FAST_ to use

All you need to do is,
1. download it from [here](https://github.com/lzan98/ip/releases/tag/A-Jar)
2. double click it
3. add your tasks
4. let it manage your tasks for you. 😉 😛 

And it's **FREE!**

Features

- [x] Manage Tasks
- [ ] Set Reminders (coming soon!)
- [ ] Managing Deadlines (coming soon!)

##

If you are a Java programmer, you can use it to practise Java too. Here is the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

lzan98 committed:
Level-1: Greet, Echo, Exit

lzan98 committed:
Level 2. Add, List

lzan98 committed:
Level-3: Mark as Done

lzan98 committed:
Level-4: ToDo, Event, Deadline

lzan98 committed:
A-TextUiTesting: Automated Text UI Testing

lzan98 committed:
Level-5: Handle Errors

lzan98 committed:
Level-6: Delete

lzan98 committed:
Level-7: Save

lzan98 committed:
Level-8: Dates and Times

lzan98 committed:
Level-7: Save (v2)

lzan98 committed:
Level-8: Dates and Times (v2)

lzan98 committed:
Merge branch 'branch-Level-7'

lzan98 committed:
Level-8: Dates and Times

lzan98 committed:
Modify output format for tasks

lzan98 committed:
Modify output for Event and Deadline

lzan98 committed:
Modify output format for tasks

lzan98 committed:
Merge branch 'branch-Level-8'

lzan98 committed:
Created more OOP: TaskList and Storage classes

lzan98 committed:
Create more OOP: Ui and Parser classes

lzan98 committed:
Add packages Duke and Tasks

lzan98 committed:
Add JUnit Tests for Duke, TaskList and Parser

lzan98 committed:
Add JavaDocs comments

lzan98 committed:
Update code to follow Coding Standard

lzan98 committed:
Merge branch 'branch-A-JavaDoc'

lzan98 committed:
Merge branch 'branch-A-CodingStandard'

lzan98 committed:
Add command to find tasks

lzan98 committed:
Merge branch 'branch-Level-9'

lzan98 committed:
Merge branch 'master' into branch-Level-9

lzan98 committed:
Edit module from ip to main

lzan98 committed:
Change program logic to save tasks to hard disk after every command

lzan98 committed:
Change text to be printed for find command

lzan98 committed:
Merge branch 'branch-Level-9'

lzan98 committed:
Shift Duke into duke package

lzan98 committed:
Merge branch 'add-gradle-support' into master

lzan98 committed:
Add gradle support

lzan98 committed:
Add and run checkstyle using Gradle

lzan98 committed:
Add a GUI to Duke

onefangg commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,158 @@
+        new Duke("Data/tasks.txt").run();
+    }*/
+
+    public String handleCommand(String command) {
```
Mildly offtrack but method has about 4-5 levels of indentations - is it possible to abstract the logic here into smaller methods?

lzan98 replied:
I agree - thanks for pointing it out! I will try to either abstract or make use of switch statements.

onefangg commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,76 @@
+     * @param filePath Path of file containing the tasks stored in hard disk.
+     */
+    public Storage(String filePath) {
+        f = new File(filePath);
```
Could the variable here be changed to a noun eg. `file` instead?

onefangg commented on src/main/java/tasks/Task.java:
```diff
@@ -0,0 +1,42 @@
+        this.description = description;
+        this.isDone = false;
+    }
+
```
I like the getters and setters defined neatly here

onefangg commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,158 @@
+                        s = s + t;
+                    }
+                }
+                return s;
```
What is the `s` variable referring to here?

onefangg commented on src/main/java/tasks/TaskList.java:
```diff
@@ -0,0 +1,67 @@
+        try {
+            Scanner s = new Scanner(f);
+            while (s.hasNext()) {
+                String str = s.nextLine();
+                String[] strings = str.split(",");
```
What is the difference between naming str vs strings?

onefangg commented on src/main/java/tasks/Todo.java:
```diff
@@ -0,0 +1,44 @@
+     * @param blean A Boolean value set to True to differentiate the creation of task from saved list and user input.
+     * @throws DukeException If the format is not followed or there are missing information.
+     */
+    public Todo(String saveFormat, boolean blean) throws DukeException {
```
What does blean stand for?

onefangg reviewed (commented):
It looks good - only some slight refactoring needed

lzan98 committed:
Fix minor bugs to GUI and update JavaDocs.

lzan98 committed:
Fix bugs with find and reorganise code for readability.

charmainehly commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,49 @@
+    public String parse(String str) throws MissingDescriptionException, DukeException, InvalidTaskException {
+        String[] strArr = str.split(" ");
+        String firstWord = strArr[0];
+        if (firstWord.equals("deadline") || firstWord.equals("event") || firstWord.equals("todo")
```
Perhaps you could also consider placing these possible 'first words' into a private String[], and utilise the contains() method to check if the firstWord from the argument passed matches any of the array items? This might possibly help to avoid having to repeat similar lines of code in the conditional statement! :)

lzan98 replied:
Yes I didn't think of that. Thank you!

charmainehly commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,76 @@
+import tasks.Task;
+import tasks.TaskList;
+
+public class Storage implements StorageInterface {
```
Perhaps JavaDoc header comments could be added for the classes as well?

charmainehly commented on src/main/java/tasks/Deadline.java:
```diff
@@ -0,0 +1,66 @@
+    }
+
+    /**
+     * Creates a Deadline task from previously saved list of tasks during initialisation of chat-bot.
```
I like the level of description in the JavaDoc comments for the methods! Perhaps you could also consider using @link to reference another member of the java classes!

charmainehly commented on src/main/java/tasks/Todo.java:
```diff
@@ -0,0 +1,44 @@
+
+import duke.DukeException;
+
+public class Todo extends Task {
```
This is minor point - perhaps you could consider renaming the class `Todo` to `ToDo` (Pascal Case) to follow the coding standards?

lzan98 replied:
Oh my - I completely missed this out. 😅 Thank you for pointing it out!

charmainehly reviewed (commented):
I think your JavaDoc comments for the methods are clear and descriptive! :) I left some minor comments for your review!

lzan98 committed:
Fix bug where not all tasks are listed

lzan98 committed:
Merge branch 'branch-Level-10'

lzan98 committed:
Remove redundant files

lzan98 committed:
Merge branch 'branch-A-Gradle'

lzan98 committed:
Rename Todo class to follow pascal naming conventions

lzan98 committed:
Implement Assertions into codebase

Codebase contains many try-catch statements as a form of extra protection
should the program not run as intended, enabling one to detect bugs.

Assertions are statement in the Java programming language that
enables one to test your assumptions about your program.

Writing assertions while programming is one of the
quickest and most effective ways to detect and correct bugs.

For a start, let's
- include assertions for switch statements with no default case.
- assert that a directory is created after calling the method mkdirs() in
the save method of Storge, removing the need for try-catch statements.

lzan98 committed:
Merge pull request #1 from lzan98/branch-A-Assertions

Implement Assertions into codebase

lzan98 committed:
Improve code quality

Duke.handleCommand has over 60 lines of code.

Among various dimensions of code quality, such as run-time efficiency,
security, and robustness, one of the most important is understandability

For a start, I have reduced the lines of code in the Duke.handleCommand
method from 60 lines to 40 lines by making my code follow more OOP.
While it might not be under 30 LOC yet, I will further attempt to
reduce it in subsequent updates.

lzan98 committed:
Merge branch 'master' into branch-A-CodeQuality

lzan98 committed:
Merge pull request #2 from lzan98/branch-A-CodeQuality

Improve code quality

lzan98 committed:
Fix bug in save format of Deadline tasks

Stauses of Deadline tasks are not saved in the proper manner,
resulting in mark and unmark commands being erased each time the
user closes the chat bot.

The bug lies in the updating of the string saveFormat during the
creation of a Deadline task, and has since been rectified.

lzan98 committed:
Implement schedule command for users to view events on specific date

lzan98 committed:
Merge branch 'branch-BCD-Extension'

lzan98 committed:
Improve Code Quality

lzan98 committed:
Refresh GUI style to better differentiate between user and bot

lzan98 committed:
Fix bug of welcome message not appearing in GUI

lzan98 committed:
Fix bug where DateTimeParseException is not being caught

lzan98 committed:
Add a representative screenshot of the product to the docs folder

lzan98 committed:
Rename Duke to Yeowoo

lzan98 committed:
Add a user guide for Yeowoo

lzan98 committed:
Update build.grade to launch from appropriate class

lzan98 committed:
Add gradle.yml file for Github CI

lzan98 committed:
Update grade.yml file

lzan98 committed:
Fix minor checkstyle violations

lzan98 committed:
Fix checkstyle violations in test

lzan98 committed:
Remove I/O Redirection in gradle.yml

lzan98 committed:
Fix bug where expected test output is incorrect

lzan98 committed:
Modify logic for ParserTest to test for exception
